sympy/solvers/solveset.py

```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    with the actual structure of the API.
+
+    4) Less complex API
+
+    The API of `\_transolve` is designed in a way that it is less
+    complex to read and understand unlike in `\_tsolve` where the
+    API structure is a mess with lots of recursive and function
+    calls making it hard to understand.
+
+    5) Faster Computation
+
+    Solving equation via `\_transolve` is much faster as compared to
+    `\_tsolve`. In `solve` attempts are made computing every possibility
+    to get the solutions. This series of attempts makes solving a bit
+    slow. Whereas in `\_transolve` computation begins only when the
+    equation is identified of being a particular type.
+
+    How to add new class of equations
+    =================================
+
+    `\_transolve` is designed in such a way that it becomes an easy task to
+    add a new class of equation solver.
+
+    The first task that needs to be done for adding your own solver
+    is to decide from where its `identification helper` will be
+    invoked. To do so determine the general form of the class of the
+    equation, for example the general form of the exponential equation
+    is `a*f(x) + b*g(x)` where `f(x)` and `g(x)` are power terms, so we
+    need to place the invocation condition inside `Add` case. Once the
+    place for `identification helper` is determined, you can add a call
+    to its `solving helper`.
+
+    For your class of equation you need to define your own
+    identification and solving helpers. The `identification helper`
+    should be implemented for generalised cases and should return
+    either `True` if the given equation belongs to the class otherwise
+    `False`. `Solving helpers` needs to be implemented with
+    heuristics or algorithms that solves majority of the equations
+    belonging to that class. The value returned could be either the
+    exact solution or a reduced form of the equation which `solveset`
+    can handle.
+
+    Apart from this, few other things needs to be taken care while adding
+    an equation solver:
+
+    - Naming conventions:
+      Name of the `identification helper` should be like  `\_check\_class`
+      where `class` will be the name or abbreviation of the class of
+      equation. The `solving helper` will be named as `\_class\_solver`.
+      For example: for exponential equation it becomes `\_check\_expo`
+      and `\_expo\_solver`.
+    - Be sure to consider corner cases.
+    - Adding tests for each helper.
+    - Adding a docstring to your helper that describes the method
+      implemented.
+      Following things needs to be included while writing the
+      documentation for the helpers:
+
+      - What is the purpose of the helper.
+      - How it solves the equation and how does it identifies the equation.
+      - Examples should be included with its proof of correctness.
+      - What are the input parameters and what does the helper returns.
```

Comment:
shouldn't the input parameters always be the same for all helpers?

Reply:
I am not sure as of now about it. By doing this we are in a way restricting helper solvers to have a limited number of parameter(s), for eg:  [_check_log](https://github.com/sympy/sympy/pull/14792/files#diff-eec0422923e8f100745c015cd8fdd6cfR1193)
won't require a second parameter `symbol` unlike `_check_expo` therefore only the `f` is passed.

Do we need to restrict input parameters?

Reply:
Every check function should get f and symbol and should only be working with the parts of the equation that contain symbol. A function can choose to ignore the symbol (as log currently does) but I would recommend against that: make it separarate args into those that depend on symbol and those that don't and only transform th esymbol-dependent parts. (Otherwise, you can run into empty recursion situations where, for example, log puts things together which then get taken apart.)